Fully qualify types in generic synthetic method return types

theron-wang opened the pull request:
Previously, when a synthetic method with a generic return type had a type parameter in the same package as the target method, it would not keep its fully qualified name in its synthetic definition, leading to compilation errors seen in na-791b.

For example, `ImmutableSet.copyOf` returned an `ImmutableSet<Feature>`, but in its synthetic definition, `copyOf` returned `com.google.common.collect.ImmutableSet<Feature>` instead of `com.google.common.collect.ImmutableSet<com.github.benmanes.caffeine.cache.Feature>`.

kelloggm commented on src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java:
```diff
@@ -3565,18 +3565,74 @@ private String lookupFQNs(String javacType) {
+      }
+    } else {
+      // If it's not imported, it's probably in the same package
+      fullyQualifiedName.append(currentPackage).append(".").append(typeArgument.toString());
```
What if the type argument was already fully-qualified in the input. For example, what if we added the following method to your test case:
```
    public Bar<com.foo.InOtherPackage2> foo() {
        return Bar.getOtherPackage2();
    }
```

I think the current implementation will hit this line and append the current package to the front of the name, resulting in something like `"com.example.com.foo.InOtherPackage2"`. This would probably be easy enough to fix by just checking if the name is fully-qualified already before you call this method.

Regardless, please add the test case.

theron-wang replied:
@kelloggm I looked into this, and unfortunately it's more complicated than that. The `lookupFQN()` method, in this case, is called after `JavaTypeCorrect`, which gets rid of the fully qualified name. Since it's not imported, I can't resolve the fully qualified name through `classAndPackageMap`. I also don't think we can simply add the types to `classAndPackageMap` in `visit(ClassOrInterfaceType)`, since there could be multiple classes with the same simple name (i.e. `com.example.Foo` and `com.other.Foo`). Do you have any ideas on how to solve this?

kelloggm replied:
In the general case I think you're right that there's no unambiguous way to handle this.

I don't think handling this case correctly is critical at the moment. Let's move the `InOtherPackage2` version of the test into a separate test runner, allow that test runner to fail intentionally, and then add a TODO in `lookupFQN` with your notes about this problem. Then we can merge this PR, which does improve the current situation even if it is imperfect.

Then, we can come back to this problem later when we're not worried about a deadline at the end of the week :)

kelloggm reviewed (changes requested):
Overall looks great, but I have one concern that needs to be addressed.

theron-wang commented:
I added another test case, as you suggested. For the expected output I put the incorrect one, but I changed the check compilation scripts to ignore the compilation error for this case.

kelloggm commented on src/test/java/org/checkerframework/specimin/MethodReturnFullyQualifiedGenericTest.java:
```diff
@@ -0,0 +1,20 @@
+ * This test checks that Specimin includes the fully qualified class name in generics when
+ * generating a synthetic method. For example, if a return type is Bar<com.foo.Foo>, the actual
+ * synthetic return type should be com.qualified.Bar<com.foo.Foo>. At this moment, we should expect
+ * this test case to fail.
```
Rather than "we should expect this test to fail", I would phrase this as "this test encodes Specimin's current behavior, which doesn't produce compilable output. If you break this test, it might not be a bad thing."

kelloggm commented on src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java:
```diff
@@ -3598,18 +3598,81 @@ private String lookupFQNs(String javacType) {
+      // TODO: handle already fully qualified generic type arguments. Right now JavaTypeCorrect
+      // only outputs the simple class name, so there is no way to get its fully qualified name
+      // here without ambiguity (i.e. org.example.Foo and com.example.Foo are different signatures)
+      // with the same simple class name
```
Add a reference to the test that demonstrates the "bad" behavior to this comment.

kelloggm reviewed (approved):
I don't need to review again after you make the requested changes to the wording of a few comments. Nice work :)

theron-wang commented:
Thanks :) Just changed the comments.
